GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -523,7 +523,27 @@ public bool TryLoad(out string error)
                 this.UpgradeRing = RingType.NoConfig;
 
                 string ringConfig = null;
-                if (this.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, out ringConfig, out error))
+                string upgradeFeedUrl = null;
+
+                // 1. Try to read ring from upgrade.feedurl first.
+                if (this.LocalConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl, out upgradeFeedUrl, out error) &&
+                    !string.IsNullOrEmpty(upgradeFeedUrl))
+                {
+                    string[] expectedRings = new string[] { RingType.Slow.ToString(), RingType.Fast.ToString() };
+                    foreach (string ring in expectedRings)
+                    {
+                        string ringSubstring = $"@{ring}/";
+                        if (upgradeFeedUrl.IndexOf(ringSubstring, StringComparison.OrdinalIgnoreCase) != -1)

```

Comment:
Can the @ symbol appear anywhere else in the URL? Is there any more checks / restrictions we should put on the URLs we consider to avoid any false hits?

I think this logic would be a great candidate to extract into its own method with its set of unit tests.

Reply:
@jrbriggs correct me if I'm wrong - it is safe to assume that `@Fast/` or `@Slow/` will not occur more that once.